File: android-client/app/src/main/java/com/cursormobile/app/data/ChatPreferencesManager.kt
package com.cursormobile.app.data

import android.content.Context
import android.content.SharedPreferences
import com.cursormobile.app.data.models.ChatMode
import com.cursormobile.app.data.models.ChatTool

/**
 * Manages persistent user preferences for new chat creation
 */
class ChatPreferencesManager(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences(
        PREFS_NAME,
        Context.MODE_PRIVATE
    )

    companion object {
        private const val PREFS_NAME = "chat_preferences"
        private const val KEY_LAST_TOOL = "last_selected_tool"
        private const val KEY_LAST_MODE = "last_selected_mode"

        @Volatile
        private var instance: ChatPreferencesManager? = null

        fun getInstance(context: Context): ChatPreferencesManager {
            return instance ?: synchronized(this) {
                instance ?: ChatPreferencesManager(context.applicationContext).also {
                    instance = it
                }
            }
        }
    }

    // MARK: - Save Preferences

    fun savePreferences(
        tool: ChatTool,
        mode: ChatMode
    ) {
        prefs.edit().apply {
            putString(KEY_LAST_TOOL, tool.value)
            putString(KEY_LAST_MODE, mode.value)
            apply()
        }
    }

    // MARK: - Load Preferences

    fun getLastTool(): ChatTool? {
        val value = prefs.getString(KEY_LAST_TOOL, null) ?: return null
        return ChatTool.entries.find { it.value == value }
    }

    fun getLastMode(): ChatMode? {
        val value = prefs.getString(KEY_LAST_MODE, null) ?: return null
        return ChatMode.entries.find { it.value == value }
    }

    // MARK: - Clear Preferences

    fun clearAllPreferences() {
        prefs.edit().apply {
            remove(KEY_LAST_TOOL)
            remove(KEY_LAST_MODE)
            apply()
        }
    }
}

File: ios-client/CursorMobile/CursorMobile/Services/ChatPreferencesManager.swift
import Foundation
import SwiftUI

/// Manages persistent user preferences for new chat creation
class ChatPreferencesManager {
    static let shared = ChatPreferencesManager()

    private let defaults = UserDefaults.standard

    // Keys for UserDefaults
    private enum Keys {
        static let lastAgentId = "lastSelectedAgentId"
        static let lastModelId = "lastSelectedModelId"
        static let lastMode = "lastSelectedMode"
        static let lastPermissionMode = "lastSelectedPermissionMode"
    }

    private init() {}

    // MARK: - Save Preferences

    func savePreferences(
        agentId: String?,
        modelId: String?,
        mode: ChatMode,
        permissionMode: PermissionMode
    ) {
        if let agentId = agentId {
            defaults.set(agentId, forKey: Keys.lastAgentId)
        }
        if let modelId = modelId {
            defaults.set(modelId, forKey: Keys.lastModelId)
        }
        defaults.set(mode.rawValue, forKey: Keys.lastMode)
        defaults.set(permissionMode.rawValue, forKey: Keys.lastPermissionMode)
    }

    // MARK: - Load Preferences

    func lastAgentId() -> String? {
        return defaults.string(forKey: Keys.lastAgentId)
    }

    func lastModelId() -> String? {
        return defaults.string(forKey: Keys.lastModelId)
    }

    func lastMode() -> ChatMode? {
        guard let rawValue = defaults.string(forKey: Keys.lastMode) else {
            return nil
        }
        return ChatMode(rawValue: rawValue)
    }

    func lastPermissionMode() -> PermissionMode? {
        guard let rawValue = defaults.string(forKey: Keys.lastPermissionMode) else {
            return nil
        }
        return PermissionMode(rawValue: rawValue)
    }

    // MARK: - Clear Preferences

    func clearAllPreferences() {
        defaults.removeObject(forKey: Keys.lastAgentId)
        defaults.removeObject(forKey: Keys.lastModelId)
        defaults.removeObject(forKey: Keys.lastMode)
        defaults.removeObject(forKey: Keys.lastPermissionMode)
    }
}
